library/sonar_plugin.py

```diff
@@ -0,0 +1,429 @@
+            self.data.append(data)
+
+
+def apply_post_api(sonar_url, api, module, plugin_data, status):
+    url = sonar_url + api
+    requests.post(url, data=plugin_data,
+                  auth=HTTPBasicAuth(module.params['username'],
+                                     module.params['password']))
+
+    msg = switch_msg(status)
+
+    if is_plugin_pending(plugin_data['key'], status, module):
+        module.exit_json(changed=True, stdout=msg)
+
+
+def switch_msg(status):
+    return {
+        'installing': 'SUCCESS: The plugin is installed',
+        'removing': 'SUCCESS: The plugin was removed',
+        'updating': 'SUCCESS: The plugin latest version is installed',
+    }[status]
+
+
+def get_key_by_name(sonar_url, api, module):
+    key = 'not-found'
+    url = sonar_url + api
+    response = requests.get(url, auth=HTTPBasicAuth(module.params['username'],
+                                                    module.params['password']))
+    json_obj = response.json()
+
+    for plugin in json_obj['plugins']:
+        if plugin['name'] == module.params['name']:
+            key = plugin['key']
+
+    return key
+
+
+def is_plugin_pending(key, status, module):
+    is_pending = False
+    pending_api = '/api/plugins/pending'
+    cancel_all_api = '/api/plugins/cancel_all'
+    hostname = module.params['hostname']
+    port = module.params['sonar_port']
+
+    url = 'http://' + hostname + ':' + str(port) + pending_api
+    response = requests.get(url, auth=HTTPBasicAuth(module.params['username'],
+                                                    module.params['password']))
+    json_obj = response.json()
+
+    if "errors" in json_obj:
+        cancel_url = 'http://' + hostname + ':' + str(port) + cancel_all_api
+        requests.post(cancel_url, auth=HTTPBasicAuth(module.params['username'],
+                                                     module.params['password']))
+        module.exit_json(changed=False,
+            stdout='FAILED: This plugin is broken or not exist in repo. The operation '
+                'canceled and all plugins was removed from pending list')
+
+    if status == 'custom':
+        for stat in json_obj:
+            if stat != 'removing':
+                for curr_stat in json_obj[stat]:
+                    if curr_stat['key'] == key:
+                        is_pending = True
+    else:
+        for stat in json_obj[status]:
+            if stat['key'] == key:
+                is_pending = True
+
+    return is_pending
+
+
+def download_custom_plugin(plugin_url, download_folder, status):
+    filename = os.path.basename(plugin_url)
+
+    filename_path = download_folder + filename
+    urllib.urlretrieve(plugin_url, filename_path)
+
+    if status == 'custom':
+        key = 'not-found'
+        match = re.search('sonar-(.*)-plugin', filename)
+        if match is not None:
+            key = match.group(1)
+            final_key = re.sub('-', '', key)
+            return final_key
```

Comment:
return is not always called, move it to outside the if block

Reply:
fixed